cirq/devices/hypergraph.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import (
+        Any, Dict, FrozenSet, Hashable, Iterable, Optional, Set, Tuple)
+
+
+AdjacencyList = Set[FrozenSet[Hashable]]
+
+class UndirectedHypergraph:
+    def __init__(self,
+                 vertices: Optional[Iterable[Hashable]]=None,
+                 labelled_edges: Optional[Dict[Iterable[Hashable], Any]]=None,
+                 ) -> None:
+        """A labelled, undirected hypergraph.
+
+        Args:
+            vertices: The vertices.
+            labelled_edges: The labelled edges, as a mapping from (frozen) sets
+                of vertices to the corresponding labels. Vertices are
+                automatically added.
+        """
+
+        self._adjacency_lists = {} # type: Dict[Hashable, AdjacencyList]
```

Comment:
It might be simpler to use just the edge dictionary (i.e. an edge set representation). Singleton vertices could be represented by the edge key `{node}`.

Reply:
The convention you suggest for singletons wouldn't work because `node` could be iterable. For example, an expected use case is lattice-based graph devices in which a reasonable vertex would be `(x, y)`.

Using just the edge dictionary is a possibility, but would require more time for common actions. The design choice I made reflects a particular space-time trade-off that seemed reasonable, but I'm open to others.

Reply:
The removal of `HashQubit` obviates my objection to the singleton convention, but the point about the space-time trade-off remains relevant.